### PR TITLE
Performance improvements, documentation cleanup

### PR DESCRIPTION
**Pull request by Noeda**
Most important performance improvement in this PR is sharing of cursors in `TrailDB.trails()` call. Cursors are relatively heavy so this can be a massive improvement in some workloads.

Also cleaned up a whole bunch of documentation and added some sphinx configuration for that.

### CODE REVIEW

**Comment by Noeda** [comment]
Travis tests seem broken but tests do complete successfully when I run them locally.

File: traildb/traildb.py [hunk 1]
```diff
@@ -224,6 +229,15 @@ def __del__(self):
     def __iter__(self):
         return self
 
+    def get_trail(self, trail_id):
+        if lib.tdb_get_trail(self.cursor, trail_id) != 0:
+            raise TrailDBError("Failed to initalize trail in cursor")
+
+            if self.event_filter_obj:
+                if lib.tdb_cursor_set_event_filter(self.cursor, event_filter_obj.flt):
+                    raise TrailDBError("cursor_set_event_filter failed")
```
Comment (aholyoke) [inline traildb/traildb.py]: This indentation looks wrong. How is it possible to reach this code immediately after raising an exception? @knutin

File: traildb/traildb.py [hunk 2]
```diff
@@ -216,6 +219,8 @@ def __init__(self,
             self.event_filter_obj = event_filter_obj
             if lib.tdb_cursor_set_event_filter(cursor, event_filter_obj.flt):
                 raise TrailDBError("cursor_set_event_filter failed")
+        else:
+           self.event_filter_obj = None
```
Comment (aholyoke) [inline traildb/traildb.py]: Indentation uses 3 spaces on this line

File: traildb/traildb.py [hunk 3]
```diff
@@ -224,6 +229,15 @@ def __del__(self):
     def __iter__(self):
         return self
 
+    def get_trail(self, trail_id):
+        if lib.tdb_get_trail(self.cursor, trail_id) != 0:
+            raise TrailDBError("Failed to initalize trail in cursor")
+
+            if self.event_filter_obj:
+                if lib.tdb_cursor_set_event_filter(self.cursor, event_filter_obj.flt):
```
Comment (aholyoke) [inline traildb/traildb.py]: `event_filter_obj` should be `self.event_filter_obj`

File: traildb/traildb.py (hunk 2 above)
Comment (Noeda) [reply traildb/traildb.py]: :scream: I shall fix this.

File: traildb/traildb.py (hunk 1 above)
Comment (Noeda) [reply traildb/traildb.py]: It's probably just misindented. I'll pull that indentation back a bit.

File: traildb/traildb.py [hunk 4]
```diff
@@ -427,21 +567,22 @@ def time_range(self, parsetime=False):
...
     def create_filter(self, event_filter):
+        """:returns: :py:class:`~traildb.TrailDBEventFilter` object created from this TrailDB."""
         return TrailDBEventFilter(self, event_filter)
 
     def apply_whitelist(self, uuids):
         """
-        Apply a whitelist of the given uuids so that only
-        events with the given uuids are returned by the
-        cursor. (Empty trails are still returned for other uuids.)
+        Applies a whitelist UUIDs to TrailDB so that further calls to
```
Comment (rnowling-adroll) [inline traildb/traildb.py]: I think we're missing an `of` in this sentence

File: traildb/traildb.py [hunk 5]
```diff
@@ -464,9 +605,8 @@ def apply_whitelist(self, uuids):
 
     def apply_blacklist(self, uuids):
         """
-        Apply a blacklist of the given uuids so that
-        only events without the given uuids are returned by the
-        cursor. (Empty trails are still returned for the given uuids.)
+        Applies a blacklist UUIDs to TrailDB so that further calls to
```
Comment (rnowling-adroll) [inline traildb/traildb.py]: Same here about having an `of`

File: traildb/traildb.py [hunk 6]
```diff
@@ -275,50 +346,82 @@ def __init__(self, path):
...
 
-    def trails(self, selected_uuids=None, **kwds):
+    def trails(self, selected_uuids=None, distinct_cursors=False, **kwds):
         """
         Iterate over all trails in this TrailDB.
 
-        The selected_uuids keyword argument can be used to select a subset
-        of trails to iterate over. Missing uuids are skipped over.
+        :param selected_uuids: If passed, only go through the UUIDs passed in
+          this argument. It should be an iterable that yields hex UUIDs.
```
Comment (rnowling-adroll) [inline traildb/traildb.py]: Should we add an explicit check of each UUID as we iterate over it? (Or do we check already?)  If performance is a concern there, then I guess any bad strings just won't be matched and a trail won't be returned for them.  Maybe we need to add a comment that entries that aren't found are silently skipped and no error is raised then?

**Pull request merged**